api/godog/requests.go

```diff
@@ -0,0 +1,184 @@
+			return err
+		}
+	}
+
+	keyPattern := "{" + key + "}"
+	if !strings.Contains(api.Request.Url, keyPattern) {
+		return errors.New("api.Request path does not contain variable: " + keyPattern)
+	}
+
+	api.Request.Url = strings.ReplaceAll(api.Request.Url, keyPattern, value)
+	return
+}
+
+func (api *BaseFeature) SetRequestBodyParameterTo(key, value string) (err error) {
+	if strings.HasPrefix(value, ".") {
+		if value, err = api.GetValue(value); err != nil {
+			return err
+		}
+	}
+
+	keyParts := strings.Split(key, ".")
+	prevMap := api.Request.Body
+	for idx, key := range keyParts {
+		if len(keyParts) == idx+1 {
+			prevMap[key] = value
+			break
+		}
+
+		if _, exists := prevMap[key]; !exists {
+			prevMap[key] = make(map[string]interface{})
+		}
+		prevMap = prevMap[key].(map[string]interface{})
+	}
+
+	return
+}
+
+func (api *BaseFeature) ExecuteTheRequest() (err error) {
+	jsonBody, err := json.Marshal(api.Request.Body)
+	if err != nil {
+		return errors.Wrap(err, "json.Marshal failed")
+	}
+
+	req, err := http.NewRequest(api.Request.Method, api.Request.Url, bytes.NewBuffer(jsonBody))
+	if err != nil {
+		return errors.Wrapf(err, "http.NewRequest failed - Body: `%s`", string(jsonBody))
+	}
+
+	req.Header = api.Request.Headers
+	req.Header.Set("Content-Type", "application/json")
+
+	api.Request.ExecutionTime = time.Now()
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.Wrapf(err, "client.Do failed - header: `%+v`", req.Header)
+	}
+
+	defer resp.Body.Close()
+	if jsonBody, err = ioutil.ReadAll(resp.Body); err != nil {
+		return errors.Wrap(err, "ioutil.ReadAll failed")
+	}
+
+	api.Response.Raw = resp
+	api.Response.Body = jsonBody
+
+	return nil
+}
+
+func (api *BaseFeature) AssertNotEmpty(responseKey string) error {
+	value, err := json_matcher.Read(api.Response.Body, responseKey)
+	if err != nil {
+		return err
+	}
+
+	if value == "" {
+		return errors.New(fmt.Sprintf("No value found for: %v", responseKey))
+	}
+	return nil
+}
+
+func (api *BaseFeature) AssertResponseBodyValueEquals(key, expected string) (err error) {
+	switch key {
+	case "len(.data)":
+		return api.AssertDataLength(expected)
+	}
+
+	if expected, err = api.GetValue(expected); err != nil {
+		return
+	}
+
+	actual, err := json_matcher.Read(api.Response.Body, key)
+	if err != nil {
+		return
+	}
+
+	if actual != expected {
+		return errors.Errorf("Match error: Values mismatch, expected: '%s' actual: '%s'", expected, actual)
+	}
+
+	return
+}
+
+func (api *BaseFeature) AssertDataLength(expected string) error {
+	expectedLen, err := strconv.Atoi(expected)
+	if err != nil {
+		return err
+	}
+
+	return json_matcher.ArrayLen(api.Response.Body, ".data", expectedLen)
+}
+
+func (api *BaseFeature) AssertResponseCode(code int) (err error) {
+	if api.Response.Raw.StatusCode != code {
+		err = errors.Errorf("wrong status code in Response: %d, Response: %s, Request: %+v", api.Response.Raw.StatusCode, string(api.Response.Body), api.Request)
```

Comment:
Should the error messages be more of testing context? I.e. "...expected: 666 actual: 777..."